File: backend/src/models/calligraphy.rs
use serde::{Deserialize, Serialize};
use sqlx::types::ipnetwork::IpNetwork;
use sqlx::FromRow;
use time::OffsetDateTime;
use uuid::Uuid;

/**
 * 書き初めデータ
 * DBのcalligraphiesテーブルに対応するモデル
 */
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Calligraphy {
  /// ユーザーID
  pub user_id: Uuid,
  /// ユーザー名
  pub user_name: String,
  /// 書き初め内容
  pub content: String,

  /// IPアドレス (情報収集用)
  #[serde(skip)]
  pub ip_address: Option<IpNetwork>,
  /// User-Agent (情報収集用)
  #[serde(skip)]
  pub user_agent: Option<String>,
  /// Accept-Language (情報収集用)
  #[serde(skip)]
  pub accept_language: Option<String>,

  // JSONシリアライズ時はISO8601形式にするための指定。
  #[serde(with = "time::serde::iso8601")]
  pub created_at: OffsetDateTime,

  #[serde(with = "time::serde::iso8601")]
  pub updated_at: OffsetDateTime,
}

// --- DTOs (Data Transfer Objects) ---
/// フロントから受け取る書き初め作成・更新用のリクエストボディ
#[derive(Deserialize)]
pub struct CreateCalligraphyRequest {
  pub user_name: String,
  pub content: String,
}

/// APIレスポンス用のDTO
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CalligraphyResponse {
  pub user_name: String,
  pub content: String,
  #[serde(with = "time::serde::iso8601")]
  pub created_at: OffsetDateTime,
  #[serde(with = "time::serde::iso8601")]
  pub updated_at: OffsetDateTime,
  pub is_mine: bool,
}

File: frontend/src/types/calligraphy.ts
/**
 * 書き初めデータの型定義
 */
export interface Calligraphy {
	user_id?: string; // 一意のID（keyとして使用）
	user_name: string;
	content: string;
	created_at: string;
	updated_at: string;
	is_mine: boolean;
}

/**
 * 書き初め作成リクエストの型定義
 */
export interface CreateCalligraphyRequest {
	user_name: string;
	content: string;
}

/**
 * APIエラーの型定義
 */
export interface ApiError {
	error: string;
	code?: string;
	details?: Record<string, unknown>;
}

/**
 * APIエラーレスポンスの型定義
 */
export interface ApiErrorResponse {
	error: string;
	code?: string;
	details?: Record<string, unknown>;
}

